src/lib.rs

```diff
@@ -72,34 +75,56 @@
 //! # }
 //! ```
 
+#![cfg_attr(not(feature = "std"), no_std)]
 #![deny(missing_docs, missing_debug_implementations)]
 
-use failure::Fail;
 use rand_core::{CryptoRng, RngCore};
 use serde_json::Error as JsonError;
 
-use std::{fmt, marker::PhantomData};
+use core::{fmt, marker::PhantomData};
 
 mod cipher_with_mac;
 mod erased;
 mod traits;
 mod utils;
 
+// Polyfill for `alloc` types.
+mod alloc {
+    #[cfg(not(feature = "std"))]
+    extern crate alloc;
+
+    #[cfg(not(feature = "std"))]
+    pub use alloc::{
+        borrow::ToOwned, boxed::Box, collections::BTreeMap, string::String, vec, vec::Vec,
+    };
+    #[cfg(feature = "std")]
+    pub use std::{
+        borrow::ToOwned, boxed::Box, collections::BTreeMap, string::String, vec, vec::Vec,
+    };
+}
+
 // Crypto backends.
+#[cfg(feature = "pure")]
+pub mod pure;
```

Comment:
Do I understand correct that we need this module if we compile with `no_std` only, right ? In this case maybe you should replace `#[cfg(feature = "pure")]` with #[cfg(not(feature = "std"))]` ?